### ISSUE TITLE
Ways to mount folder inside host machine to auxiliary machine?

### ISSUE THREAD

**Issue by soanni** [codenvy/codenvy#499]
Hello, Codenvy Team.

Is there any possible way to share the /projects folder inside workspace machine to auxiliary machine?
The picture is like that:

1. I have a workspace with a project source code. It is mounted from host to workspace /projects/<name_of_project> folder. 
2. There is a ready docker image with all configured services (for example TomCat)
3. I can create in current workspace additional "treatment" machine form that image
4. I made some development on the code base. Then i built it and had new .war file as a result.
5. I need a way to inject this file to the "treatment" machine's TomCat webapps folder in order to check how my app behaves after my development.
6. I am a regular user of the system and i don't have access to SSH codenvy host.

As far as i understand workspace machine is run like that

docker run - v /projects ...

How can i run the additional treatment machine the same way?

Thank you.

**Comment by TylerJewell** [comment]
Yes - there is a way.  But you need to have a small sense of adventure.

Your workspace dev machine has sshd running inside of it. It is possible to use sshfs from within the auxiliary machine to connect to the workspace machine and then mount a specific `/projects` folder within the auxiliary machine.  We have a form of this sophistication built into an experimental CLI for Che that we have, though it is more complicated than what you need because it also uses the unison framework, which you will not need.

You'd then need to make sure that the public / private key-pair between the dev & auxiliary machine is properly configured so that sshfs can work directly.

**Comment by eivantsov** [comment]
@soanni another option is to use rsync

**Comment by eivantsov** [comment]
@soanni let me know if you need help with this - I have a few ready to use commands that you can try right away.

**Comment by soanni** [comment]
@eivantsov thank you. i'm interested in everything new regarding Codenvy so i'm ready to try your commands.

**Comment by MuhamadSherif** [comment]
@eivantsov would you please provide these commands you have as I am now facing issue with this, I need to mount my container with a NAS server.

**Comment by eivantsov** [comment]
@MuhamadSherif I do not think rsync commands will solve mounting problem. Perhaps, if you elaborate on the use case, I can be more explicit.

**Comment by TylerJewell** [comment]
@eivantsov @JamesDrummond  - we are getting this question quite a bit, so its maybe time for us to start to prepare a list of ways to mount external folders into a workspace.  This is my starting list. This may not be accurate or complete. 

1: Install your own Codenvy Enterprise. Within CE, we give admins the ability to mount additional host directories into users workspaces. So this would be a situation where you have a closed environment and host mounting is allowed.

2: You can sshfs mount a workspace directory in `/projects` into another system using sshfs. We use this in the `che mount` facility today.  Can link on how this is possible.  This then requires the person who performs the mount to then do a copy / move / rsync exercise.  This is a scenario where the external system is reaching into the workspace.

3: You can setup a unison command, which does a recurring synchronization between a workspace and another external drive. This also requires ssh access to the workspace.  This is a scenario where the external system is reaching into the workspace.

4: In scenarios where the workspace wants to push out - then the full range of FTP / SFTP / rsynch / SSH commands from one system to another is available.

**Comment by tolusha** [comment]
@TylerJewell If we allow admin to mount extra volumes It will have effect per installation. Each user will have the same mounted directories.

**Comment by tolusha** [comment]
IMHO it make sense for CHE only.

**Comment by TylerJewell** [comment]
Yeah - we understand the risk and limits of this.  There are some advantages temporarily though.

**Comment by tolusha** [comment]
So, user story will be:
As a admin I should able to mount the same volumes for all machines for all users.

**Comment by TylerJewell** [comment]
Yes.

**Comment by bmicklea** [comment]
@tolusha @gazarenkov In speaking with the customer the following would be sufficient:

- Allow Codenvy admins to define a set of available mounts for workspaces, they enter information necessary to connect to the volume (obviously it has to be reachable from the codenvy instance)
- Codenvy checks to ensure they're not mounting /etc or /bin
- Users would be able to reference any of those mount points in their workspaces / compose / factory definitions
- The syntax might look something like the following (you can suggest alternatives):

`{ "machineName: "ws-main",`
`  ....,`
`  ....,`
`   "mountPoints":[`
`    { `
`     "mountPoint":"/folder1",`
`      "alias": "/data"`
`    },`
`    { `
`      "mountPoint":"/folder2",`
`      "alias":"/shared"`
`    } `
`  ]`
`},`
`{`
` "machineName: "machine-oracle",`
`   ....,`
`   ....,`
`   "mountPoints":[`
`    { `
`     "mountPoint":"/folder1",`
`      "alias": "/logs"`
`    }`
`  ]`
`}`

- For the first phase mounts would not be available to choose from the user dashboard

The ideal solution has been captured in a separate issue: https://github.com/codenvy/codenvy/issues/709

**Comment by tolusha** [comment]
New requirements are different from ones taken into accounts before estimating the issue.
This issue supposes to mount folder through a property into configuration, for intance:
`machine.docker.machine_extra_volumes`

**Comment by bmicklea** [comment]
Yes, the definition of the mount folder can happen as a property in configuration. That's fine. No UI here.

**Comment by tolusha** [comment]
And no extra configuration inside workspace / factory / compose definitions

**Comment by gazarenkov** [comment]
@bmicklea I do not understand what is this syntax about at all.
As discussed: Could you please collect user stories to understand the purpose (format of those stories does not matter). We do not expect implementations details (configuration format, UI, model etc), we need to understand what do they need at the end to estimate if and what kind of workaround we can do.
The directions we want to understand arer:
- what kind of information they want to store on external volumes (files, db, config etc ...real life cases would be very helpful here)
- is this information supposed to be persistent - i.e. is it expected that it is not changed/removed when we stop containers which use it
- is it shared between users/workspaces
- other details...

**Comment by bmicklea** [comment]
I just spoke to @soanni to get a few more details @gazarenkov you can ask questions to him as well if needed.

**Their context:**
- They want to limit developer SSH access to machines as much as possible.
- Only DevOps should be able to setup mounts for workspaces in order to avoid potential mounting problems.
- All mounted directories should have their contents persisted when the container is stopped.

**Their use cases:**
(1) Reconfiguring services quickly and consistently
 - A development team has a set of configuration files for their services (one or more for each version of each service) held in a central directory outside the workspace.
 - This directory is mounted into the machine that runs the service.
 - The developer can quickly change the running configuration of the service by copying the various configuration files into the appropriate location in their workspace.
 - This makes it very fast and easy for developers to test their code against service A-1.2 and service B-5.3, then switch to service A-1.3 and service B-4.9.

(2) Moving built binaries between workspace machines
 - The application build will be executed on the Codenvy workspace's devmachine, but it may need to be moved to one of the other machines in the workspace.
 - Having a volume mount is one way to solve this.

Today they also use mounts to facilitate sharing of logs and other files with the team but with Codenvy that will be more easily done by sharing access to the developer's specific workspace.

**Comment by garagatyi** [comment]
5.0 has a volumes_from feature in compose environment type. I suppose it might help in that case. Unfortunately this feature doesn't work due to bug #2366

**Comment by bmicklea** [comment]
I'm still not clear on why volumes_from is critical here.

From what I have read, volumes_from is just a way to allow a container to use a set of volumes already defined in another container. It seems like a shortcut to having to re-define volumes in each container of a compose environment. But I thought that in M1 we weren't going to be able to do the volume mount at all - is that wrong?

What I'm trying to understand is that if we give 5.0.0-M1 to a customer can they try it and expect to be able to mount volumes in Codenvy?

**Comment by tolusha** [comment]
Postponed to 5.0.0-M2

**Comment by bmicklea** [comment]
@tolusha - the che issue (https://github.com/eclipse/che/issues/2366) may get pushed to M2, but so far it's not clear to me that the Che issue is the full solution to this issue. I need more information or explanation.

**Comment by JamesDrummond** [comment]
#728 Extra volume mounting.

**Comment by tolusha** [comment]
`machine.server.extra.volume` has been introduces. It is a way to mount extra volumes into docker container.

For instance:
```
machine.server.extra.volume=/home/user/folder1:/mnt/folder1:ro,Z;/home/user/folder2:/mnt/folder2:ro,Z
```

**Comment by bmicklea** [comment]
@tolusha I believe we still need the fix for https://github.com/eclipse/che/issues/2366 before we can close this. Your PR alone doesn't really let us do the second use case smoothly.

**Comment by tolusha** [comment]
They are two separated issues.

**Comment by bmicklea** [comment]
I understand that your PR and the Che issue are separate and cover different use cases but this issue is about the two use cases outlined here: https://github.com/codenvy/codenvy/issues/499#issuecomment-245941445

So I don't want to close this issue until both are complete and smooth.

**Comment by bmicklea** [comment]
Adjusted milestone to match the release where VOLUMES_FROM will be fixed.

**Comment by bmicklea** [comment]
Closing issue as https://github.com/eclipse/che/issues/2366 has been closed in M4 as well.
